### ISSUE TITLE
fetch/update a single bug?

### ISSUE THREAD

**Issue by srl295** [ralphbean/bugwarrior#946]
may be related to #516 somewhat

Maybe I should just be more patient. 

However, sometimes there's a new bug assigned to me or an important change to a bug. It would be nice to be able to request a partial pull, of just one specific bug:
- to pick up a new assignment
- to mark it as closed
- to pick up some important change to description or label

I use bugwarrior every day with all of my clients. It could be helpful to shorten the time to switch to a new task.

Some of the queries take a while, and have rate limits. So a shorter query could reduce server load for everyone.

**Comment by ryneeverett** [comment]
Yes, once #516 is implemented for a given service it ought to be fairly easy to add a command to just pull a given issue through the same method and feed it into the synchronization.

However I'm doubtful there would be enough interest to justify this feature and it seems like resources would be better spent on economizing requests and improving the way we work with rate limits.

I don't know if this is even a thing but it would be nice if services offered an api parameter to filter out issues that haven't changed since *X* timestamp. Then one should be able to greatly increase the polling frequency without threatening to surpass rate limits.

**Comment by ryneeverett** [comment]
Having used the configuration option added in #898 a few times, I do not find the design ergonomic. I don't really want to edit my configuration file to add a single bug.

I'd rather have a subcommand like `bugwarrior fetch <url>` and I think this could potentially resolve the original concern in this issue as well if we added a flag like `--no-track` to tell bugwarrior to fetch the bug without concerning itself about tracking it separately in the future since it's already been pulled with the current configuration.

The immediate problem is that for my use case where I do want it tracked, without adding the url to the configuration bugwarrior wouldn't know to update it and would in fact close it on the next `bugwarrior pull` because it wouldn't be returned. I can think of a couple approaches to get around this:

1. Use BugwarriorData. As far as I recall, we only use the `bugwarrior.data` file we store in ~/.task to cache tokens for bitbucket. We could also add fetched url's here and add them to the list of things to sync on each pull, just like the `github.issue_urls` parameter currently does. This would be easy, but the downside is that the `bugwarrior.data` file would be of more permanent value. It's kind of nice that it's effectively just a cache file now and could be deleted or forgotten when you change computers without consequence.
2. Change the synchronization logic. We could mark tracked issues with a UDA like BUGWARRIOR_TRACK_ISSUE=true and when the issue is encountered during synchronization it could be updated rather than closed. I hate to add any more logic to the sprawling and critical synchronization function, but it doesn't seem like it would take a whole lot of code. I'm liking this idea better at the moment.

Another issue is that a configured service would have to be selected for each fetch. I'm thinking there could be an optional `--service <service-config-name>` parameter but that the service could usually be inferred and the user could be prompted to make a selection in the case of ambiguity.

From an implementation standpoint, another issue is that the way I implemented #898 doesn't scale well and other services might have a more complex translation between issue url's and api url's. A better approach would be to put the logic in the base class and only add to the service API grammars to translate from identifiers into an Issue. Perhaps this would be something like a regex with an arbitrary number of named matchers and a method that takes those names as arguments and returns an Issue. Pseudocode:

```py

class GithubService(Service):
    ISSUE_URL = [
        r"https://github.com/(?P<owner>.*)/(?P<repo>.*)/issues/(?P<number>[0-9]*)"  # full url
        r"(?P<owner>.*)/(?P<repo>.*)#(?P<number>[0-9]*)"  # short identifier
    ]

    def fetch_issue(self, owner, repo, number) -> GithubIssue:
        ...
```
